### PR TITLE
postponed days for Russian holidays in 2024

### PR DESCRIPTION
**Pull request by summicron502**
## Proposed change
Added postponed days for Russian holidays in 2024

Added days and translations

## Type of change

- [ ] New country/market holidays support (thank you!)
- [x] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've run `make pre-commit`, it didn't generate any changes
- [x] I've run `make test`, all tests passed locally

[contributing-guidelines]: https://github.com/vacanza/python-holidays/blob/dev/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/python-holidays/tree/dev/docs/source

### CODE REVIEW

File: holidays/countries/russia.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
 from gettext import gettext as tr
 
-from holidays.calendars.gregorian import JAN, FEB, MAY
+from holidays.calendars.gregorian import JAN, FEB, APR, MAY, DEC
```
Comment (PPsyrius) [inline holidays/countries/russia.py:15]:
```suggestion
from holidays.calendars.gregorian import JAN, FEB, APR, MAY, JUN, DEC
```

File: holidays/countries/russia.py [hunk 2]
```diff
@@ -120,4 +120,13 @@ class RussiaStaticHolidays:
...
             (MAY, 8, tr("День Победы")),
         ),
+        # Postponed days 2024
+        # src: https://www.consultant.ru/document/cons_doc_LAW_455140/
+        2024: (
+            (APR, 29, tr("Перенос 27.04.2024")),
+            (APR, 30, tr("Перенос 02.11.2024")),
+            (MAY, 10, tr("Перенос 06.01.2024")),
+            (DEC, 30, tr("Перенос 28.12.2024")),
+            (DEC, 31, tr("Перенос 07.01.2024")),
```
Comment (PPsyrius) [inline holidays/countries/russia.py:130]:
```suggestion
            (APR, 29, APR, 27),
            (APR, 30, FEB, 11),
            (MAY, 10, JUN, 1),
            (DEC, 30, DEC, 28),
            (DEC, 31, JAN, 7),
```

File: holidays/locale/ru/LC_MESSAGES/RU.po [hunk 3]
```diff
@@ -80,3 +80,18 @@ msgstr ""
...
+msgstr ""
+
+msgid "Перенос 06.01.2024"
+msgstr ""
+
+msgid "Перенос 28.12.2024"
+msgstr ""
+
+msgid "Перенос 07.01.2024"
+msgstr ""
```
Comment (PPsyrius) [inline holidays/locale/ru/LC_MESSAGES/RU.po:97]:
```suggestion
msgid "%d.%m.%Y"
msgstr ""

msgid "Перенос %s"
msgstr ""
```

File: holidays/locale/en_US/LC_MESSAGES/RU.po [hunk 4]
```diff
@@ -82,3 +82,18 @@ msgstr "Day of consent and reconciliation"
...
+msgstr "Postpone of 02.11.2024"
+
+msgid "Перенос 06.01.2024"
+msgstr "Postpone of 06.01.2024"
+
+msgid "Перенос 28.12.2024"
+msgstr "Postpone of 28.12.2024"
+
+msgid "Перенос 07.01.2024"
+msgstr "Postpone of 07.01.2024"
```
Comment (PPsyrius) [inline holidays/locale/en_US/LC_MESSAGES/RU.po:99]:
```suggestion
msgid "%d.%m.%Y"
msgstr "%m/%d/%Y"

msgid "Перенос %s"
msgstr "Postpone of %s"
```

**Review by PPsyrius — CHANGES_REQUESTED** [review]
Thank you for your contribution, although you might want to consider switching to the dedicated substituted holidays pattern instead, as shown in #1613 . :+1:

That and some additional test cases for these postponed holidays would make this PR perfect for merging.

File: holidays/countries/russia.py [hunk 5]
```diff
@@ -120,4 +120,13 @@ class RussiaStaticHolidays:
             (FEB, 24, tr("День защитника Отечества")),
             (MAY, 8, tr("День Победы")),
         ),
+        # Postponed days 2024
+        # src: https://www.consultant.ru/document/cons_doc_LAW_455140/
```
Comment (PPsyrius) [inline holidays/countries/russia.py:124]:
```suggestion
        # src: https://www.consultant.ru/document/cons_doc_LAW_455140/

        # Date format (see strftime() Format Codes).
        substituted_date_format = tr("%d.%m.%Y")
        # Postpone of %s.
        substituted_label = tr("Перенос %s")
```

File: holidays/countries/russia.py (hunk 2 above)
Comment (PPsyrius) [reply holidays/countries/russia.py:130]: Make sure to include the following section below Line 115, too (after the `class RussiaStaticHolidays:` line)

```python3
        # Date format (see strftime() Format Codes).
        substituted_date_format = tr("%d.%m.%Y")
        # Postpone of %s.
        substituted_label = tr("Перенос %s")
```

**Comment by summicron502** [comment]
Changes accepted

File: holidays/locale/ru/LC_MESSAGES/RU.po [hunk 6]
```diff
@@ -80,3 +80,9 @@ msgstr ""
 #. Anniversary of the Great October Socialist Revolution.
 msgid "Годовщина Великой Октябрьской социалистической революции"
 msgstr ""
+
+msgid "%d.%m.%Y"
+msgstr ""
+
+msgid "Перенос %s"
+msgstr ""
```
Comment (PPsyrius) [inline holidays/locale/ru/LC_MESSAGES/RU.po:88]:
```suggestion
#. Date format (see strftime() Format Codes).
msgid "%d.%m.%Y"
msgstr ""

#. Postpone of %s.
#, c-format
msgid "Перенос %s"
msgstr ""
```

File: holidays/locale/en_US/LC_MESSAGES/RU.po [hunk 7]
```diff
@@ -82,3 +82,9 @@ msgstr "Day of consent and reconciliation"
 #. Anniversary of the Great October Socialist Revolution.
 msgid "Годовщина Великой Октябрьской социалистической революции"
 msgstr "Anniversary of the Great October Socialist Revolution"
+
+msgid "%d.%m.%Y"
+msgstr "%m/%d/%Y"
+
+msgid "Перенос %s"
+msgstr "Postpone of %s"
```
Comment (PPsyrius) [inline holidays/locale/en_US/LC_MESSAGES/RU.po:90]:
```suggestion
#. Date format (see strftime() Format Codes).
msgid "%d.%m.%Y"
msgstr "%m/%d/%Y"

#. Postpone of %s.
#, c-format
msgid "Перенос %s"
msgstr "Postpone of %s"
```

File: holidays/locale/ru/LC_MESSAGES/RU.po [hunk 8]
```diff
@@ -14,7 +14,7 @@
 #
 msgid ""
 msgstr ""
-"Project-Id-Version: Python Holidays 0.29\n"
+"Project-Id-Version: Python Holidays 0.47\n"
```
Comment (PPsyrius) [inline holidays/locale/ru/LC_MESSAGES/RU.po:17]:
```suggestion
"Project-Id-Version: Python Holidays 0.48\n"
```

File: holidays/locale/en_US/LC_MESSAGES/RU.po [hunk 9]
```diff
@@ -14,7 +14,7 @@
 #
 msgid ""
 msgstr ""
-"Project-Id-Version: Python Holidays 0.29\n"
+"Project-Id-Version: Python Holidays 0.47\n"
```
Comment (PPsyrius) [inline holidays/locale/en_US/LC_MESSAGES/RU.po:17]:
```suggestion
"Project-Id-Version: Python Holidays 0.48\n"
```

**Review by PPsyrius — CHANGES_REQUESTED** [review]
@summicron502 Other changes (I've put them as subcomments so they may have been overlooked before):

`holidays/countries/russia.py` Line 115-137 (`class RussiaStaticHolidays` section)
```python
class RussiaStaticHolidays:
    # Date format (see strftime() Format Codes).
    substituted_date_format = tr("%d.%m.%Y")
    # Postpone of %s.
    substituted_label = tr("Перенос %s")

    special_public_holidays = {
        # Postponed Days 2023
        # src: https://www.consultant.ru/document/cons_doc_LAW_425407/
        2023: (
            (FEB, 24, JAN, 1),
            (MAY, 8, JAN, 8),
        ),
        # Postponed Days 2024
        # src: https://www.consultant.ru/document/cons_doc_LAW_455140/
        2024: (
            (APR, 29, APR, 27),
            (APR, 30, FEB, 11),
            (MAY, 10, JUN, 1),
            (DEC, 30, DEC, 28),
            (DEC, 31, JAN, 7),
        ),
    }
```
`tests\countries\test_russia.py` Line 30-40 (`def test_special_holidays(self)` section)
```python
    def test_special_holidays(self):
        self.assertHoliday(
            # Postponed Holidays.
            "2023-02-24",
            "2023-05-08",
            "2024-04-29",
            "2024-04-30",
            "2024-05-10",
            "2024-12-30",
            "2024-12-31",
        )
```

**Comment by arkid15r** [comment]
It's just a friendly reminder to address the CI/CD issues before we could merge this PR.

Thank you!

**Comment by PPsyrius** [comment]
> @summicron502 Other changes (I've put them as subcomments so they may have been overlooked before):
> 
> `holidays/countries/russia.py` Line 115-137 (`class RussiaStaticHolidays` section)
> 
> ```python
> class RussiaStaticHolidays:
>     # Date format (see strftime() Format Codes).
>     substituted_date_format = tr("%d.%m.%Y")
>     # Postpone of %s.
>     substituted_label = tr("Перенос %s")
> 
>     special_public_holidays = {
>         # Postponed Days 2023
>         # src: https://www.consultant.ru/document/cons_doc_LAW_425407/
>         2023: (
>             (FEB, 24, JAN, 1),
>             (MAY, 8, JAN, 8),
>         ),
>         # Postponed Days 2024
>         # src: https://www.consultant.ru/document/cons_doc_LAW_455140/
>         2024: (
>             (APR, 29, APR, 27),
>             (APR, 30, FEB, 11),
>             (MAY, 10, JUN, 1),
>             (DEC, 30, DEC, 28),
>             (DEC, 31, JAN, 7),
>         ),
>     }
> ```
> 
> `tests/countries/test_russia.py` Line 30-40 (`def test_special_holidays(self)` section)
> 
> ```python
>     def test_special_holidays(self):
>         self.assertHoliday(
>             # Postponed Holidays.
>             "2023-02-24",
>             "2023-05-08",
>             "2024-04-29",
>             "2024-04-30",
>             "2024-05-10",
>             "2024-12-30",
>             "2024-12-31",
>         )
> ```

Please make sure to include this part manually since GitHub doesn't allow code reviewer to suggest sections where the original PR contributor doesn't touch.

File: holidays/locale/en_US/LC_MESSAGES/RU.po [hunk 10]
```diff
@@ -82,3 +82,12 @@ msgstr "Day of consent and reconciliation"
...
 msgstr "Anniversary of the Great October Socialist Revolution"
+
+#. Date format (see strftime() Format Codes).
+msgid "%d.%m.%Y"
+msgstr "%m/%d/%Y"
+
+#. Postpone of %s.
+#, c-format
+msgid "Перенос %s"
+msgstr "Postpone of %s"
```
Comment (arkid15r) [inline holidays/locale/en_US/LC_MESSAGES/RU.po:93]: We normally use  `Выходной (перенесено с %s)` `Day off (substituted from %s)` for the labels

**Review by arkid15r — COMMENTED** [review]
Please find my comments to address below:

**Comment by arkid15r** [comment]
> For some reason l10n started to fail
> 
> > from lingva.extract import main as create_pot_file
> > ModuleNotFoundError: No module named 'lingva'
> 
> Is it known problem or that's my fault?

try running `make setup` or `pip install -U lingva`

**Comment by arkid15r** [comment]
@summicron502 did you have a chance to work on addressing the issues/comments?

**Pull request closed without merging**